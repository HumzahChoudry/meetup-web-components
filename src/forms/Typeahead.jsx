import PropTypes from 'prop-types';
import React from 'react';
import Downshift from 'downshift';
import cx from 'classnames';

import TextInput from './TextInput';

export const TA_DROPDOWN_CLASSNAME = 'typeahead-dropdown';
export const TA_ITEM_CLASSNAME = 'typeahead-item';

const defaultItemToString = itemValue => (typeof itemValue === 'string' ? itemValue : '');

/**
 * @module Typeahead
 */
class Typeahead extends React.PureComponent {
	constructor(props) {
		super(props);

		this.handleSelection = this.handleSelection.bind(this);
		// this.handleSingleSelection = this.handleSingleSelection.bind(this);
		// this.handleMultiselectSelect = this.handleMultiselectSelect.bind(this);
		this.stateReducer = this.stateReducer.bind(this);
	}

	// handleMultiselectSelect(item) {
	// 	this.props.onSelect(item, [...this.props.multiSelectValues, item]);
	// }

	// handleSingleSelection(selectedItem, stateAndHelpers) {
	// 	!this.props.multiSelect && this.props.onSelect && this.props.onSelect(selectedItem, stateAndHelpers);
	// }

	handleSelection(selectedItem, stateAndHelpers) {
		if(this.props.multiSelect) {
			this.props.onSelect(selectedItem, [...this.props.multiSelectValues, selectedItem]);
		} else {
			this.props.onSelect && this.props.onSelect(selectedItem, stateAndHelpers);
		}
	}

	stateReducer(state, changes) {
		switch (changes.type) {
			case Downshift.stateChangeTypes.keyDownEnter:
			case Downshift.stateChangeTypes.clickItem:
				this.handleSelection(changes.selectedItem);

				return {
					...changes,
					selectedItem: this.props.multiSelectValues ? this.props.multiSelectValues : changes.selectedItem,
					isOpen: this.props.openOnFocus,
				};

			case Downshift.stateChangeTypes.blurInput:
				return {
					...changes,
					isOpen: this.props.openOnFocus,
				};

			default:
				return changes;
		}
	}

	render() {
		const {
			inputProps,
			height,
			items,
			multiSelect,
			multiSelectValues,
			openOnFocus,
			itemToString,
			...other
		} = this.props;

		delete other.onSelect;

		return (
			<Downshift
				onSelect={this.handleSelection}
				selectedItem={multiSelectValues}
				stateReducer={openOnFocus && this.stateReducer}
				itemToString={itemToString}
				{...other}
			>
				{({
					getInputProps,
					getItemProps,
					isOpen,
					highlightedIndex,
					selectedItem,
					openMenu
				}) =>
				(<div className="typeahead">
					<TextInput
						{...getInputProps({
							...inputProps,
							className: 'typeahead-input',
							onFocus: openOnFocus && openMenu
						})}
					/>
					{Boolean(isOpen && items && items.length)
						&&
							<div
								className={TA_DROPDOWN_CLASSNAME}
								style={height && {
									height: height,
									overflowY: 'scroll'
								}}
								>
								{items &&
									items.map((item, i) => {
										const selected = multiSelect && selectedItem && selectedItem.includes(item.props.value);

										return (
											<div
												{...getItemProps({
													item: item.props.value,
													i,
													key: `typeaheadItem-${i}`,
													id: `typeaheadItem-${i}`,
													className: cx(
														TA_ITEM_CLASSNAME,
														item.props.className,
														{
															'typeahead-item--isActive': highlightedIndex == i,
														}
													),
													...item.props
												})}
											>
												{
													typeof item.props.children === 'function'
														? item.props.children({ isSelected: selected })
														: item.props.children
												}
											</div>
										);}
									)
								}
							</div>
					}
				</div>)}
			</Downshift>
		);
	}
}

Typeahead.defaultProps = {
	itemToString: defaultItemToString
};

Typeahead.propTypes = {
	inputProps: PropTypes.object,
	items: PropTypes.arrayOf(PropTypes.element),
	itemToString: PropTypes.func,
	height: PropTypes.string
	// add multiselect and openOnFocus proptypes
	// add something for onSelect that makes it required if multiSelect is true
};

export default Typeahead;
