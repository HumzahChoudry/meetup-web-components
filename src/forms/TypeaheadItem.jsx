import PropTypes from 'prop-types';

/**
 * @module TypeaheadItem
 */
const TypeaheadItem = ({value, className, children}) => children;

export default TypeaheadItem;

TypeaheadItem.propTypes = {
	value: PropTypes.oneOfType([
		PropTypes.string,
		PropTypes.array,
		PropTypes.object
	]).isRequired,
	className: PropTypes.string
	// children: PropTypes.oneOfType([
	// 	PropTypes.node,
	// 	PropTypes.function
	// ]).isRequired
};
