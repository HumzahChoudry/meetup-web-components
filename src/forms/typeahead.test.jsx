import React from 'react';
import { shallow, mount } from 'enzyme';

import Typeahead, {
	TA_DROPDOWN_CLASSNAME,
	TA_ITEM_CLASSNAME
} from './Typeahead';

import TypeaheadItem from './TypeaheadItem';

const HEIGHT_LIMIT = '100px';
const INPUT_NAME = 'testName';
const TA_ITEMS = [
	<TypeaheadItem value="Item One">
		<div>
			<h3 className="text--bold">Item One</h3>
			<p>Is super cool</p>
		</div>
	</TypeaheadItem>,
	<TypeaheadItem value="Item Two">
		<div>
			<h3 className="text--bold">Item Two</h3>
			<p>Is super cool</p>
		</div>
	</TypeaheadItem>,
	<TypeaheadItem value="Item Three">
		<div>
			<h3 className="text--bold">Item Three</h3>
			<p>Is super cool</p>
		</div>
	</TypeaheadItem>,
	<TypeaheadItem value="Item Four">
		<div>
			<h3 className="text--bold">Item Four</h3>
			<p>Is super cool</p>
		</div>
	</TypeaheadItem>
];

describe('Typeahead', () => {
	const closedComponent = mount(
		<Typeahead inputProps={{name: INPUT_NAME}} />);

	const openComponent = mount(
		<Typeahead
			isOpen
			items={TA_ITEMS}
			height={HEIGHT_LIMIT}
			inputProps={{name: INPUT_NAME}}
		/>);

	const dropdownArea = openComponent.find(`.${TA_DROPDOWN_CLASSNAME}`);

	it('should hide dropdown area by default', () => {
		const dropdownArea = closedComponent.find(`.${TA_DROPDOWN_CLASSNAME}`);
		expect(dropdownArea.length).toBe(0);
	});

	it('should show dropdown area when isOpen is true', () => {
		expect(dropdownArea.length).toBe(1);
	});

	it('should crop dropdown area content when a height is provided', () => {
		expect(dropdownArea.prop('style').height).toBe(HEIGHT_LIMIT);
	});

	it('should pass props to the TextInput component', () => {
		const textInput = closedComponent.find('input');
		expect(textInput.prop('name')).toBe(INPUT_NAME);
	});

	it('should render `items` passed in', () => {
		expect(dropdownArea.find(`.${TA_ITEM_CLASSNAME}`).length).toBe(TA_ITEMS.length);
	});
});

describe('TypeaheadItem', () => {
	it('exists', () => {
		const taItem = shallow(
			<TypeaheadItem
				value="lolnothing"
			>
				<p>matters</p>
			</TypeaheadItem>
		);
		expect(taItem).toMatchSnapshot();
	});
});
