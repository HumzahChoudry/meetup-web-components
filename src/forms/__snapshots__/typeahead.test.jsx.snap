// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeaheadItem exists 1`] = `
<p>
  matters
</p>
`;
